### PR TITLE
Implemented range and offset levels

### PR DESCRIPTION
**Pull request by bharath2438**
Adds the range level format and the coord_bounds and coord_access level functions.
cc @hameerabbasi

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -51,7 +51,7 @@ file(GLOB_RECURSE sources CONFIGURE_DEPENDS "${CMAKE_CURRENT_SOURCE_DIR}/source/
 
 # Note: for header-only libraries change all PUBLIC flags to INTERFACE and create an interface
 # target: add_library(${PROJECT_NAME} INTERFACE)
-add_library(${PROJECT_NAME} INTERFACE ${headers} ${sources})
+add_library(${PROJECT_NAME} INTERFACE ${headers} ${sources} "test/source/range_test.cpp")
```
Comment (hameerabbasi) [inline CMakeLists.txt:54]: This change isn't needed, we are globbing all the files anyway.

File: include/xsparse/levels/range.hpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
+#ifndef XSPARSE_LEVELS_RANGE_HPP
```
Comment (hameerabbasi) [inline include/xsparse/levels/range.hpp:1]: You need both range and offset for DIA.

Comment (hameerabbasi) [reply include/xsparse/levels/range.hpp:1]: Dense is innermost, then range, then offset.

File: include/xsparse/levels/range.hpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            range(IK size_N, IK size_M, OffsetContainer&& offset)
+                : m_size_N(std::move(size_N))
+                , m_size_M(std::move(size_M))
+                , m_offset(offset)
+            {
+            }
+
+            inline std::pair<IK, IK> coord_bounds(typename BaseTraits::I i) const noexcept
+            {
+                static_assert(std::tuple_size<i> > 1, "Tuple size should be greater than 1");
```
Comment (hameerabbasi) [inline include/xsparse/levels/range.hpp:54]:
```suggestion
                static_assert(std::tuple_size<i> >= 1, "Tuple size should be at least 1");
```

File: include/xsparse/levels/range.hpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    template <class... LowerLevels, class IK, class PK, class OffsetContainer>
+    struct util::coordinate_position_trait<
+        levels::range<std::tuple<LowerLevels...>, IK, PK, OffsetContainer>>
+    {
+        using Coordinate = IK;
+        using Position = PK;
+    };
+}
+
+#endif  // XSPARSE_LEVELS_RANGE_HPP
```
Comment (hameerabbasi) [inline include/xsparse/levels/range.hpp:82]: Newline missing.

**Review by hameerabbasi — COMMENTED** [review]
I've left some preliminary comments. Please take a look, and add tests.

**Comment by hameerabbasi** [comment]
Overall most of this looks good. We're just missing tests and the offset level.

File: include/xsparse/levels/range.hpp (hunk 4 above)
Comment (bharath2438) [reply include/xsparse/levels/range.hpp:82]: what change is to be made here?

Comment (hameerabbasi) [reply include/xsparse/levels/range.hpp:82]: As discussed, on chat, there needs to be a newline after `#endif`.

File: include/xsparse/levels/range.hpp [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                                                                  PK,
+                                                                  OffsetContainer>
+        {
+            using BaseTraits
+                = util::base_traits<range, std::tuple<LowerLevels...>, IK, PK, OffsetContainer>;
+
+        public:
+            range(IK size_N, IK size_M)
+                : m_size_N(std::move(size_N))
+                , m_size_M(std::move(size_M))
```
Comment (hameerabbasi) [inline include/xsparse/levels/range.hpp:30]:
```suggestion
                : m_size_M(std::move(size_M))
                , m_size_N(std::move(size_N))
```

For optimization reasons, all fields must be initialized in the same order in which they're declared, and declared in the same order in which they're used. Smaller fields should go on top and larger ones on the bottom.

Comment (hameerabbasi) [reply include/xsparse/levels/range.hpp:30]: This applies to all constructors and classes/structs.

**Review by hameerabbasi — CHANGES_REQUESTED** [review]
Last round of changes to get CI passing, otherwise LGTM.

File: test/source/range_test.cpp [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+                CHECK(p2 == p3);
+                CHECK(l2 + offset[l1] == i3);
+            }
+            ++l2;
+        }
+        CHECK(l2 == std::min(SIZE_N, SIZE_M - offset[l1]));
+        ++l1;
+    }
+    CHECK(l1 == SIZE);
+}
```
Comment (hameerabbasi) [inline test/source/range_test.cpp:53]: Newline here.

File: include/xsparse/levels/range.hpp (hunk 5 above)
Comment (bharath2438) [reply include/xsparse/levels/range.hpp:30]: The paper uses NxM (rows x columns) convention while referring to the dimensions of the matrix, would it be best to change the declaration and initializations to follow this convention?

File: test/source/range_test.cpp (hunk 6 above)
Comment (bharath2438) [reply test/source/range_test.cpp:53]: Thanks, will do.

File: include/xsparse/levels/range.hpp (hunk 5 above)
Comment (hameerabbasi) [reply include/xsparse/levels/range.hpp:30]: That works, yes.

**Pull request merged**